src/Umbraco.Cms.ManagementApi/Controllers/RecycleBinControllerBase.cs

```diff
@@ -0,0 +1,115 @@
+﻿using Microsoft.AspNetCore.Mvc;
+using Umbraco.Cms.Core;
+using Umbraco.Cms.Core.Models;
+using Umbraco.Cms.Core.Models.Entities;
+using Umbraco.Cms.Core.Services;
+using Umbraco.Cms.ManagementApi.ViewModels.RecycleBin;
+
+namespace Umbraco.Cms.ManagementApi.Controllers;
+
+public abstract class RecycleBinControllerBase<TItem> : Controller
+    where TItem : RecycleBinItemViewModel, new()
+{
+    private readonly IEntityService _entityService;
+    private readonly string _itemUdiType;
+
+    protected RecycleBinControllerBase(IEntityService entityService)
+    {
+        _entityService = entityService;
+        // ReSharper disable once VirtualMemberCallInConstructor
+        _itemUdiType = ItemObjectType.GetUdiType();
+    }
+
+    protected abstract UmbracoObjectTypes ItemObjectType { get; }
+
+    protected async Task<ActionResult<PagedResult<TItem>>> GetRoot(long pageNumber, int pageSize)
+    {
+        IEntitySlim[] rootEntities = GetPagedRootEntities(pageNumber, pageSize, out var totalItems);
+
+        TItem[] treeItemViewModels = MapRecycleBinViewModels(null, rootEntities);
+
+        PagedResult<TItem> result = PagedResult(treeItemViewModels, pageNumber, pageSize, totalItems);
+        return await Task.FromResult(Ok(result));
+    }
+
+    protected async Task<ActionResult<PagedResult<TItem>>> GetChildren(Guid parentKey, long pageNumber, int pageSize)
+    {
+        IEntitySlim[] children = GetPagedChildEntities(parentKey, pageNumber, pageSize, out var totalItems);
+
+        TItem[] treeItemViewModels = MapRecycleBinViewModels(parentKey, children);
+
+        PagedResult<TItem> result = PagedResult(treeItemViewModels, pageNumber, pageSize, totalItems);
+
+        return await Task.FromResult(Ok(result));
+    }
+
+    protected virtual TItem MapRecycleBinViewModel(Guid? parentKey, IEntitySlim entity)
+    {
+        if (entity == null)
+        {
+            throw new ArgumentNullException(nameof(entity));
+        }
+
+        var viewModel = new TItem
+        {
+            Icon = _itemUdiType,
+            Name = entity.Name!,
+            Key = entity.Key,
+            Type = _itemUdiType,
+            HasChildren = entity.HasChildren,
+            IsContainer = entity.IsContainer,
+            ParentKey = parentKey
+        };
+
+        return viewModel;
+    }
+
+    private IEntitySlim[] GetPagedRootEntities(long pageNumber, int pageSize, out long totalItems)
+    {
+        // TODO: EntityService needs to be expanded to get root identifier (key?) by UmbracoObjectType - for now this is a workaround
+        int rootId = ItemObjectType == UmbracoObjectTypes.Document
+            ? Constants.System.RecycleBinContent
+            : ItemObjectType == UmbracoObjectTypes.Media
+                ? Constants.System.RecycleBinMedia
+                : throw new ArgumentOutOfRangeException(nameof(ItemObjectType), "Only Document and Media are supported");
```

Comment:
```suggestion
        int rootId = ItemObjectType switch
        {
            UmbracoObjectTypes.Document => Constants.System.RecycleBinContent,
            UmbracoObjectTypes.Media => Constants.System.RecycleBinMedia,
            _ => throw new ArgumentOutOfRangeException(nameof(ItemObjectType), "Only Document and Media are supported")
        };
```

This might be more readable as a switch statement